akka-actor-typed-tests/src/test/java/jdocs/akka/typed/InteractionPatternsTest.java

```diff
@@ -314,6 +341,219 @@ public void timers() throws Exception {
+    }
+  }
+
+  public static Behavior<DaveProtocol> daveBehavior(final ActorRef<HalCommand> hal) {
+    return Behaviors.deferred((ActorContext<DaveProtocol> ctx) -> {
+
+      // asking someone requires a timeout, if the timeout hits without response
+      // the ask is failed with a TimeoutException
+      final Timeout timeout = Timeout.apply(3, TimeUnit.SECONDS);
+
+      ctx.ask(
+        HalResponse.class,
+        hal,
+        timeout,
+        // construct the outgoing message
+        (ActorRef<HalResponse> ref) -> new OpenThePodBayDoorsPlease(ref),
+        // adapt the response (or failure to respond)
+        (response, throwable) -> {
+          if (response != null) {
+            return new AdaptedResponse(response.message);
+          } else {
+            return new AdaptedResponse("Request failed");
+          }
+      });
+
+      // we can also tie in request context into an interaction, it is safe to look at
+      // actor internal state from the transformation function, but remember that it may have
+      // changed at the time the response arrives and the transformation is done, best is to
+      // use immutable state we have closed over like here.
+      final int requestId = 1;
+      ctx.ask(
+        HalResponse.class,
+        hal,
+        timeout,
+        // construct the outgoing message
+        (ActorRef<HalResponse> ref) -> new OpenThePodBayDoorsPlease(ref),
+        // adapt the response (or failure to respond)
+        (response, throwable) -> {
+          if (response != null) {
+            return new AdaptedResponse(requestId + ": " + response.message);
+          } else {
+            return new AdaptedResponse(requestId + ": Request failed");
+          }
+        });
+
+      return Behaviors.immutable(DaveProtocol.class)
+        // the adapted message ends up being processed like any other
+        // message sent to the actor
+        .onMessage(AdaptedResponse.class, (innerCtx, response) -> {
+          innerCtx.getLog().info("Got response from HAL: {}", response.message);
+          return Behaviors.same();
+        }).build();
+    });
+  }
+  // #actor-ask
+
+
+  // #standalone-ask
+  interface CookieCommand {}
+  static class GiveMeCookies implements CookieCommand {
+    public final ActorRef<Cookies> cookies;
+    GiveMeCookies(ActorRef<Cookies> cookies) {
+      this.cookies = cookies;
+    }
+  };
+  static class Cookies {}
+
+
+  public void askAndPrint(ActorSystem<Object> system, ActorRef<CookieCommand> cookieActorRef) {
+    CompletionStage<Cookies> result = AskPattern.ask(
+      cookieActorRef,
+      GiveMeCookies::new,
+      Timeout.apply(3, TimeUnit.SECONDS),
+      system.scheduler());
+
+    result.whenComplete((cookies, failure) -> {
+      if (cookies != null) System.out.println("Yay, cookies!");
+      else System.out.println("Boo! didn't get cookies in time.");
+    });
+  }
+  // #standalone-ask
+
+
+  // hidden stuffs for the per-session-child sample
```

Comment:
but, isn't it confusing to show types that are unknown?

Reply:
Hmm maybe, my thinking was that the types are not really ever used as anything but just types, they have no fields etc. In scala for example I defined them as empty case classes to not have to match on `.type` on case objects etc.

Reply:
ok, let's rebase and merge